lib/cusolver/cusolverRF.jl

```diff
@@ -0,0 +1,429 @@
+    valsA::Vector{T}
+    nnzL::Cint
+    rowsL::Vector{Cint}
+    colsL::Vector{Cint}
+    valsL::Vector{T}
+    nnzU::Cint
+    rowsU::Vector{Cint}
+    colsU::Vector{Cint}
+    valsU::Vector{T}
+    P::Vector{Cint}
+    Q::Vector{Cint}
+end
+
+function RfHostLU(
+    A::CuSparseMatrixCSR{T};
+    ordering=:AMD, tol=1e-8, check=true,
+) where T
+    m, n = size(A)
+    @assert m == n # only squared matrices are supported
+    nnzA = nnz(A)
+
+    # Transfer data to host
+    h_rowsA = A.rowPtr |> Vector{Cint}
+    h_colsA = A.colVal |> Vector{Cint}
+    h_valsA = A.nzVal |> Vector{T}
+
+    # cusolverRf is 0-based
+    h_rowsA .-= Cint(1)
+    h_colsA .-= Cint(1)
+    h_Qreorder = zeros(Cint, n)
+    # Create duplicate matrix for reordering
+    h_rowsB = copy(h_rowsA)
+    h_colsB = copy(h_colsA)
+    h_valsB = copy(h_valsA)
+
+    spH = sparse_handle()
+
+    # Create matrix descriptor
+    desca = CUSPARSE.CuMatrixDescriptor()
+    CUSPARSE.cusparseSetMatType(desca, CUSPARSE.CUSPARSE_MATRIX_TYPE_GENERAL)
+    CUSPARSE.cusparseSetMatIndexBase(desca, CUSPARSE.CUSPARSE_INDEX_BASE_ZERO)
+
+    # Reordering
+    if ordering == :AMD
+        cusolverSpXcsrsymamdHost(
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, h_Qreorder,
+        )
+    elseif ordering == :MDQ
+        cusolverSpXcsrsymmdqHost(
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, h_Qreorder,
+        )
+    elseif ordering == :METIS
+        cusolverSpXcsrmetisndHost(
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, C_NULL, h_Qreorder,
+        )
+    elseif ordering == :RCM
+        cusolverSpXcsrsymrcmHost(
+            spH,
+            n, nnzA, desca,
+            h_rowsA, h_colsA, h_Qreorder,
+        )
+    end
+
+    h_mapBfromA = zeros(Cint, nnzA)
+    @inbounds for i in 1:nnzA
+        h_mapBfromA[i] = i # identity matrix
+    end
+
+    # Compute permutation in two steps
+    size_perm = Ref{Csize_t}(0)
+    cusolverSpXcsrperm_bufferSizeHost(
+        spH,
+        m, n, nnzA, desca,
+        h_rowsB, h_colsB, h_Qreorder, h_Qreorder,
+        size_perm,
+    )
+
+    buffer_cpu = Base.Libc.malloc(size_perm[] * sizeof(Cint))
+    cusolverSpXcsrpermHost(
+        spH,
+        m, n, nnzA, desca,
+        h_rowsB, h_colsB, h_Qreorder, h_Qreorder, h_mapBfromA,
+        buffer_cpu,
+    )
+
+    # Apply permutation
+    h_valsB = h_valsA[h_mapBfromA]
+
+    # LU Factorization
+    info = Ref{CUSOLVER.csrqrInfo_t}()
+    cusolverSpCreateCsrluInfoHost(info)
+
+    cusolverSpXcsrluAnalysisHost(
+        spH,
+        m, nnzA, desca,
+        h_rowsB, h_colsB, info[],
+    )
+
+    size_internal = Ref{Cint}(0)
+    size_lu = Ref{Cint}(0)
+    cusolverSpDcsrluBufferInfoHost(
+        spH,
+        n, nnzA, desca,
+        h_valsB, h_rowsB, h_colsB,
+        info[],
+        size_internal, size_lu
+    )
+
+    n_bytes = size_lu[] * sizeof(Cint)
+    buffer_lu = Base.Libc.malloc(n_bytes)
```

Comment:
Why raw `malloc` calls and not a regular `Array`? That way you can also lose the byte calculation.

Reply:
Good catch. I have replaced `malloc` by a regular `Array`.